File: 907-koko-eating-bananas/koko-eating-bananas.cpp
#include <vector>
#include <numeric>
#include <cmath>   
using namespace std;

class Solution {
public:
    double totalHours(vector<int> &piles, int hour){
        double total_hours = 0;
        for(int i=0; i<piles.size();i++){
            total_hours += ceil((double)piles[i]/hour);
        }
        return total_hours;
    }
    
    int minEatingSpeed(vector<int>& piles, int h) {
        int n = *max_element(piles.begin(), piles.end());
        int start = 1, end = n;
        int result = end;
        while(start<=end){
            int mid = (start+end)/2;
            double total_hours = totalHours(piles, mid);
            if(total_hours <= (double)h) {
                result = mid;
                end = mid-1;
            }
            else start = mid+1;
        }
        return (int)result;
    }
};

File: 907-koko-eating-bananas/koko-eating-bananas.java
class Solution {
    public int minEatingSpeed(int[] piles, int h) {
        int low = 1;
        int high = getMax(piles);
        int ans = high;

        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (canEatAll(piles, h, mid)) {
                ans = mid;
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }

        return ans;
    }

    private boolean canEatAll(int[] piles, int h, int speed) {
        int time = 0;
        for (int pile : piles) {
            time += (pile + speed - 1) / speed; 
            if (time > h) return false;
        }
        return true;
    }

    private int getMax(int[] piles) {
        int max = piles[0];
        for (int pile : piles) {
            if (pile > max) max = pile;
        }
        return max;
    }
}
